core/git.go

```diff
@@ -283,22 +273,88 @@ func (repo *RemoteGitRepository) setup(ctx context.Context) (_ *gitutil.GitCLI,
 	if err != nil {
 		return nil, nil, err
 	}
+	var resolvPath string
 	if netConf != nil {
-		resolvPath, err := mountResolv(netConf)
+		resolvPath, err = mountResolv(netConf)
 		if err != nil {
 			return nil, nil, err
 		}
-		opts = append(opts, gitutil.WithExec(func(ctx context.Context, cmd *exec.Cmd) error {
-			return runWithStandardUmaskAndNetOverride(ctx, cmd, "", resolvPath)
-		}))
-		cleanups.Add("remove updated /etc/resolv", func() error {
-			return os.Remove(resolvPath)
-		})
+		cleanups.Add("remove updated /etc/resolv", func() error { return os.Remove(resolvPath) })
 	}
 
+	// Inject auth headers and DNS config into git commands.
+	// Auth: Uses GIT_CONFIG_* env vars to set http.extraheader for submodules.
+	// DNS: Binds custom resolv.conf in mount namespace for session consistency.
+	opts = append(opts, gitutil.WithExec(func(ctx context.Context, cmd *exec.Cmd) error {
+		if headerLine != "" {
+			base := repo.URL.Scheme + "://" + repo.URL.Host
```

Comment:
Yeah, this is the main part of the patch, I kinda missed on the first pass that this is where the magic happens.

I'm a bit... on the fence. It's kinda crappy, but it's also definitely better than what was there before.

For example, if I have a gitlab submodule in my github repo, then there's actually no way for me to specify how to clone that submodule. Which is annoying. We need better submodule APIs, but this SGTM in the meantime.

Reply:
Good point, I have no idea on how to fix it entirely though ... Would love to have your insight

Reply:
https://github.com/dagger/dagger/issues/10863